lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3521,8 +3546,23 @@ bool tryToReplaceCBVec4ArrayToScalarArray(Value *V, Type *TyV, Value *Src,
   Type *EltTy = AT->getElementType();
   unsigned sizeInBits = DL.getTypeSizeInBits(EltTy);
   // Convert array of float4 to array of float.
-  replaceScalarArrayWithVectorArray(V, Src, MC, sizeInBits >> 5);
-  return true;
+  if (replaceScalarArrayWithVectorArray(V, Src, MC, sizeInBits >> 5)) {
+    Value *DstBC = MC->getArgOperand(0);
+    MC->setArgOperand(0, UndefValue::get(MC->getArgOperand(0)->getType()));
+    DXASSERT(DstBC->user_empty(), "memcpy bitcast used by something else?");
+    if (DstBC->user_empty()) {
```

Comment:
Assert and then check the same thing feels strange.
Is it because DXASSERT only work for debug build?

Reply:
I wouldn't expect the bitcast to be used by anything else, but if it is and we are on a release build, we don't want to make the IR invalid.  Do you have a better pattern to suggest for this?

Reply:
Isn't the setArgOperand on line 3551 destructive, so that the IR is invalid here?

Wondering why this isn't an internal error?

Reply:
The setArgOperand clears the use in the replaced memcpy so it will be cleaned up as dead code later.  The idea is to get rid of any (now) dead users of `Dst` (or `V`) so we may delete it after the replacement.

We don't expect other users, but the presence of other users wouldn't represent invalid IR.  So, this asserts if the unexpected happens (other users), but still handles it properly (when asserts are disabled) by avoiding modification of the bitcast operand if there are other users, so it doesn't invalidate the IR for them.

If one of the asserts goes off, it's an indication that local assumptions may be wrong, but IR could still be perfectly valid.  Perhaps I should simply remove the asserts, but I'd be interested seeing the scenario which causes either these asserts to fire.

Reply:
I don't feel strongly enough to block on it.  But...my 2c is that if you really want to be sure you hear about any cases where the assumption is wrong then relying on asserts won't help.

Reply:
Would this be more appropriate?
```diff
   // Convert array of float4 to array of float.
   if (replaceScalarArrayWithVectorArray(V, Src, MC, sizeInBits >> 5)) {
     Value *DstBC = MC->getArgOperand(0);
     MC->setArgOperand(0, UndefValue::get(MC->getArgOperand(0)->getType()));
-    DXASSERT(DstBC->user_empty(), "memcpy bitcast used by something else?");
     if (DstBC->user_empty()) {
       // Replacement won't include the memcpy dest.  Now remove that use.
       if (BitCastInst *BCI = dyn_cast<BitCastInst>(DstBC)) {
         Value *Dst = BCI->getOperand(0);
         Type *DstTy = Dst->getType();
-        DXASSERT(Dst == V, "otherwise unexpected dest");
         if (Dst == V)
           BCI->setOperand(0, UndefValue::get(DstTy));
+        else
+          llvm_unreachable("Unexpected dest of memcpy.")
       }
+    } else {
+      llvm_unreachable("Unexpected users of memcpy bitcast.");
     }
     return true;
   }
   return false;
```

Reply:
I've pushed that change, and will resolve this comment.  Please comment again if there are still concerns.